### PR TITLE
Support for Shadow DOM

### PR DESCRIPTION
**Pull request by lmvco**
This PR fixes the usage of the `monaco-editor-core` in shadowDOM (see this issue to get more details - https://github.com/microsoft/monaco-editor/issues/308). 
This request is based on some changes proposed in https://github.com/paulvarache/monaco-editor-element. 

This PR assumes that, when the monaco-editor-core runs in a shadow DOM, the shadow root should be assigned to `window.monacoShadowRoot`. It is a workaround to make it visible in all the classes/helpers that deal with the DOM (e.g. dom.ts). I would like to know how can we attach it to a Monaco instance instead of having it global.

Lastly, since some browsers do not implement all the functions needed for the ShadowDOM, I had to add some polyfills. They were added to `editorDom.ts` (I am not sure if it's the most suitable place)

### CODE REVIEW

File: src/vs/base/browser/dom.ts [hunk 1]
```diff
@@ -820,11 +820,20 @@ export function hasParentWithClass(node: HTMLElement, clazz: string, stopAtClazz
 	return !!findParentWithClass(node, clazz, stopAtClazzOrNode);
 }
 
-export function createStyleSheet(container: HTMLElement = document.getElementsByTagName('head')[0]): HTMLStyleElement {
+export function createStyleSheet(container: HTMLElement | null = null): HTMLStyleElement {
+	if (!container) {
+		if ((window as any).monacoShadowRoot) {
```
Comment (alexdima) [inline src/vs/base/browser/dom.ts]: There can be multiple editors instantiated in a single page, so using a global here is not possible.

File: src/vs/editor/browser/controller/mouseTarget.ts [hunk 2]
```diff
@@ -836,7 +836,10 @@ export class MouseTargetFactory {
 
 	private static _actualDoHitTestWithCaretRangeFromPoint(ctx: HitTestContext, coords: ClientCoordinates): IHitTestResult {
 
-		const range: Range = document.caretRangeFromPoint(coords.clientX, coords.clientY);
+		const docOrRoot = (window as any).monacoShadowRoot || document;
```
Comment (alexdima) [inline src/vs/editor/browser/controller/mouseTarget.ts]: There can be multiple editors instantiated in a single page, so using a global here is not possible.

File: src/vs/editor/browser/editorDom.ts [hunk 3]
```diff
@@ -205,3 +205,84 @@ export class GlobalEditorMouseMoveMonitor extends Disposable {
 		});
 	}
 }
+
+
+if (typeof ShadowRoot.prototype.caretRangeFromPoint === 'undefined') {
+
+	ShadowRoot.prototype.caretRangeFromPoint = function (x, y) {
```
Comment (alexdima) [inline src/vs/editor/browser/editorDom.ts]: The editor should not add methods to the DOM elements. i.e. if someone embeds the editor on their page, there should be no such side-effects.

File: src/vs/editor/browser/editorDom.ts [hunk 4]
```diff
@@ -205,3 +205,84 @@ export class GlobalEditorMouseMoveMonitor extends Disposable {
...
+			range.setStart((el as any).firstChild, offset);
+			range.setEnd((el as any).firstChild, offset);
+		}
+
+		return range;
+	};
+}
+
+(function (window) {
+	(window as any)._getCharWidth = (char: any, font: any) => {
```
Comment (alexdima) [inline src/vs/editor/browser/editorDom.ts]: This should be done with a singleton, not with a global.

**Review by alexdima — CHANGES_REQUESTED** [review]
The editor should avoid creating globals or changing the prototype of globals. We need to figure out how to remove these hacks before merging.

File: src/vs/base/browser/dom.ts (hunk 1 above)
Comment (alexdima) [reply src/vs/base/browser/dom.ts]: I have reworked this, such that all calls to createStyleSheet pass in a relevant container in the shadow dom case.

File: src/vs/editor/browser/controller/mouseTarget.ts (hunk 2 above)
Comment (alexdima) [reply src/vs/editor/browser/controller/mouseTarget.ts]: I have reworked this such that the dom node of the editor is checked to be in the shadow DOM or not dynamically.

File: src/vs/editor/browser/editorDom.ts (hunk 4 above)
Comment (alexdima) [reply src/vs/editor/browser/editorDom.ts]: I have reworked this to be a singleton.

File: src/vs/editor/browser/editorDom.ts (hunk 3 above)
Comment (alexdima) [reply src/vs/editor/browser/editorDom.ts]: I have reworked this to avoid changing the ShadowRoot prototype.

**Comment by alexdima** [comment]
Thank you! :heart:

**Comment by tejasrivastav** [comment]
@lmvco Thanks for the contribution.
I have one question though.
when [editor](https://github.com/microsoft/vscode/blob/master/src/vs/editor/standalone/browser/standaloneCodeEditor.ts#L200) is initialized it sets the aria container on the `document.body`.

Would that cause any accessibility issue ?

**Pull request merged**